designs/odcr.md

```diff
@@ -0,0 +1,140 @@
+
+We won't introduce any other prioritization nor fallback when launching nodes against EC2NodeClasses with a Capacity Reservation. This means that it is the user's responsibility to ensure the Capacity Reservation associated to a EC2NodeClasses will include instance types, availability zone and platform the EC2NodeClasses should launch.
+This also means that we will not allow an EC2NodeClass to create Spot instances when user specified a Capacity Reservation. 
+
+Because of this this we will skip [checkODFallback](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/instance/instance.go#L200C14-L200C29) during processing.
+
+In addition, the NodeClass with Capacity Reservations will pin the NodeClass into all instance types and availability zones the Capacity Reservations reserved.
+
+_Note that these aren't permanent restrictions but simply narrowing down what features exist in the first iteration of supporting Capacity Reservation_ 
+
+Pros: 
+- This puts the onus of checking on the user to verify capatbility of their NodeClasses and requirements. It makes no assumption about what a Capacity Reservation can do nor implies can't do (Open to checking if Capacity is at limit prior to calling CreateFleet).
+- This forces users who wish to leverage fallback and prioritization to use NodePool weights or Capacity Reservation Group rather than relying on EC2NodeClass.
+
+Cons:
+- This implementation is overly restrictive causing it to be difficult to use. Its possible that the restrictions makes it too difficult for users to use EC2NodeClasses effectively.
+
+#### Pricing and consolidation
+
+
+##### Provisioning
+
+Pricing isn't directly(?) considered during provisioning, the logic is offloaded to CreateFleet. This will remain the same as Capacity Reservation will either be used and fails, if limit is hit, or it will fall back to on demand nodes that CreateFleet chooses based on on demand allocation strategy. 
+
+##### Consolidating Capacity Reserved Instances
+
+During consolidation pricing does matter as it affects which candidate will be [prioritized](https://github.com/kubernetes-sigs/karpenter/blob/75826eb51589e546fffb594bfefa91f3850e6c82/pkg/controllers/disruption/consolidation.go#L156). Since all capacity instances are paid ahead of time, their cost is already incurred. Users would likely want to prioritize filling their capacition
+reservation first then fall back into other instances. Because of this reserved instances should likely show up as 0 dollar pricing when we calculate the instance pricing. Since each candidate is tied to a NodePool and NodeClass, we should be able to safely override the pricing per node under a capacity reservation.
+
+##### Consolidating into Capacity Reserved Instances
+
+If we track Capacity Reservation usage, we can optimize the cluster configuration by moving non-Capacity Reserved instances into 
```

Comment:
I think this should just work out-of-the-box if the user has prioritization set-up with weights on their NodePools and we model the pricing correctly. I'd rather just assume that this works and then discuss what the interaction is going to look like.

Reply:
Based on some of the usage patterns we've seen so far - a form of priority and fairness needs to be considered. We don't want one nodepool, or a few nodepools, consuming the majority of the reservation and leaving other nodepools starved for resources because they are unable to provision instances

Reply:
What do you mean "move"? Do you mean between the NodePools? Because if the ODCR is open, it will automatically bind even if you didn't target it.

Reply:
I think this meant multiple node pool has the same targetted ODCR

Reply:
> We don't want one nodepool, or a few nodepools, consuming the majority of the reservation

It's an interesting question. If you have multiple clusters sharing a reservation, do you need to set limits on how far they can get into using the reservation? Are there limits that we need to prescribe to the number of instances you can consume on a given reservation that you select on? 

@bryantbiggs When you have seen users share reservations with targeted or open ODCRs, how do users ensure they separate out the reservation instance capacity fairly?

Reply:
organizations that are provided capacity via a targeted ODCR - they can have multiple teams that want/need access to some/all of that capacity. today, its very manual in terms of coordinating and bookeeping the usage across multiple teams to ensure fair and equal(ish) access